lib/puppet/state_machine/machine_builder.rb

```diff
@@ -0,0 +1,182 @@
+require 'puppet/state_machine/state'
+
+module Puppet
+  class StateMachine
+    # The MachineBuilder class provides a DSL for implementing the states and transitions of a state machine,
+    # and validating that the machine is well behaved.
+    class MachineBuilder
+      # @param machine_name [String] The name of the machine to build.
+      def initialize(machine_name)
+        @machine_name = machine_name
+        @states = {}
+        @start_state = nil
+      end
+
+      # Build and verify a new state machine.
+      # @return [Puppet::StateMachine]
+      def build(&blk)
+        yield self
+        verify!
+        Puppet::StateMachine.new(@machine_name, @states, @start_state)
+      end
+
+      # Compose multiple state machines into a a single state machine.
+      #
+      # @example
+      #   machine.compose({machine1: m1, machine2: m2}) do |m|
+      #     m.start_state([:m1, :start])
+      #
+      #     m.transition(
+      #       source: [:m1, :final],
+      #       target: [:m2, :start],
+      #       on: Puppet::StateMachine::FINAL_EVENT)
+      #
+      #     m.final([:m2, :final])
+      #   end
+      def compose(machines, &blk)
+        machines.each_pair do |ns, machine|
+          @states.merge!(machine.namespaced_states(ns))
+        end
+        build(&blk)
+      end
+
+      # Verify that the machine is well defined.
+      # @return [void]
+      def verify!
+        verify_start_state!
+        verify_transition_targets!
+        verify_exiting_transitions!
```

Comment:
Again, comment might have been eaten. Yes, these are suffixed as such since those methods exist to either silently return or throw an exception. I'm happy to remove them if this isn't consistent with the rest of the codebase.